python/GafferSceneUI/SetEditor.py

```diff
@@ -0,0 +1,379 @@
+		GafferUI.PathFilterWidget.__init__( self, self.__patternWidget, pathFilter )
+
+		self.__patternWidget._qtWidget().setPlaceholderText( "Filter..." )
+
+		self.__patternWidget.editingFinishedSignal().connect( Gaffer.WeakMethod( self.__patternEditingFinished ), scoped = False )
+		self.__patternWidget.dragEnterSignal().connectFront( Gaffer.WeakMethod( self.__dragEnter ), scoped = False )
+		self.__patternWidget.dragLeaveSignal().connectFront( Gaffer.WeakMethod( self.__dragLeave ), scoped = False )
+		self.__patternWidget.dropSignal().connectFront( Gaffer.WeakMethod( self.__drop ), scoped = False )
+
+		self.__context = None
+		self.__scene = None
+
+		self._updateFromPathFilter()
+
+	def setContext( self, context ) :
+
+		self.__context = context
+
+	def setScene( self, scene ) :
+
+		self.__scene = scene
+
+	def _updateFromPathFilter( self ) :
+
+		self.__patternWidget.setText( self.pathFilter().getMatchPattern() )
+
+	def __patternEditingFinished( self, widget ) :
+
+		self.pathFilter().setMatchPattern( self.__patternWidget.getText() )
+
+	def __dragEnter( self, widget, event ) :
+
+		self.__patternWidget.setHighlighted( True )
+
+		return True
+
+	def __dragLeave( self, widget, event ) :
+
+		self.__patternWidget.setHighlighted( False )
+
+		return True
+
+	def __drop( self, widget, event ) :
+
+		if isinstance( event.data, IECore.StringVectorData ) and len( event.data ) > 0 :
+			if event.data[0].startswith( "/" ) :
+				# treat as paths, return all sets those paths are members of
+				setNames = self.__getSetNamesFromPaths( event.data )
+			else :
+				# treat as set names
+				setNames = event.data
+
+			self.pathFilter().setMatchPattern( " ".join( sorted( setNames ) ) )
+
+		self.__patternWidget.setHighlighted( False )
+
+		return True
+
+	def __getSetNamesFromPaths( self, paths ) :
+
+		if self.__scene is None or self.__context is None :
+			return []
+
+		matchingSets = []
+		pathMatcher = IECore.PathMatcher( paths )
+		with self.__context :
+			for s in self.__scene["setNames"].getValue() :
+				if not self.__scene.set( s ).value.intersection( pathMatcher ).isEmpty() :
+					matchingSets.append( str( s ) )
+
+		return matchingSets
+
+##########################################################################
+# _EmptySetFilterWidget
+##########################################################################
+
+class _EmptySetFilterWidget( GafferUI.Widget ) :
+
+	def __init__( self, emptySetFilter ) :
+
+		self.__enabledWidget = GafferUI.BoolWidget()
+		self.__enabledWidget.setText( "Hide Empty" )
+		self.__enabledWidget.setToolTip( "Hide sets with no members." )
+
+		GafferUI.Widget.__init__( self, self.__enabledWidget )
+
+		self.__emptySetFilter = emptySetFilter
+		self.__enabledWidget.stateChangedSignal().connect( Gaffer.WeakMethod( self.__enabledStateChanged ), scoped = False )
+
+	def __enabledStateChanged( self, widget ) :
+
+		self.__emptySetFilter.setEnabled( self.__enabledWidget.getState() )
```

Comment:
Could we just use a `BasicPathFilterWidget` here instead of this custom class? If we do need a custom class, it should probably still be implemented as a subclass of `PathFilterWidget`, and update `__enabledWidget` in `_updatefromPathFilter()`.

Reply:
Swapped over to the BasicPathFilterWidget in 007d5c1. Is this Ok to create directly, or am I better calling `PathFilterWidget.create()`?

Reply:
Fine to create directly.